cpp/src/arrow/dataset/file_base.h

```diff
@@ -343,6 +343,18 @@ class ARROW_DS_EXPORT FileWriter {
   fs::FileLocator destination_locator_;
 };
 
+/// \brief Controls what happens if files exist in an output directory during a dataset
+/// write
+enum ExistingDataBehavior : int8_t {
+  /// Deletes all files in a directory the first time that directory is encountered
+  kDeleteMatchingPartitions,
+  /// Ignores existing files, overwriting any that happen to have the same name as an
+  /// output file
+  kOverwriteOrIgnore,
+  /// Returns an error if there are any files or subdirectories in the output directory
+  kError,
```

Comment:
Is this about the whole root dir? Or only the (sub)directory where actually is being written to? 

(I am thinking about the use case of a "strict" append (only adding new partitions, eg adding new days for dataset partitioned by year/month/day), where you want to error for writing to a subdirectory that already exists, but not necessarily if some files/directories are already present in the root directory)

Reply:
Ah, it seems that I asked something similar when I first look at it .. (https://github.com/apache/arrow/pull/10955#discussion_r692070088), but the API changes a bit since then.

Reply:
Unfortunately, this still has the same logic, which is a bit coarse.  It will give an error on the case you describe because it only checks to see if any files exist in the root directory.

If we check for subdirectories we either have to check on the fly or do two passes.

If we check on the fly then it could lead to case where data is partially written when an error occurs.

Two passes would be pretty inefficient and we'd probably need to spill to disk to avoid exploding memory so it's too complex for us at the moment.

Maybe someday in the future we could have a rollback mechanism so we can check on the fly and then rollback when finished.  Or we can write to a temporary directory and do a `mv` onto the destination but that wouldn't be supported on S3.

Reply:
@jorisvandenbossche Do you want to create a JIRA for this issue?

Reply:
I think it could be useful behaviour? So then yes I can create a JIRA for it.

> If we check for subdirectories we either have to check on the fly or do two passes.  
> If we check on the fly then it could lead to case where data is partially written when an error occurs.

What I don't fully understand is why it is not possible to check that efficiently on the fly in this case, while the `kDeleteMatchingPartitions` also needs to check if a directory already exists to delete its content if necessary (and also needs to do it only the first time when writing, and not for subsequent files to that directory)

Reply:
Ah, because `kDeleteMatchingPartitions` deletes data and doesn't abort the operation. While for erroring in this case, you could have already written a file to a partition that didn't exist yet, before writing a file to an existing partition. In which case you indeed expect to roll back the part that already was written.

Reply:
(and you actually also already clearly explained that in August .. https://github.com/apache/arrow/pull/10955#discussion_r694423159 Sorry! :))

Reply:
Yes, you described the problem well.  The rollback would be needed.  I created ARROW-14175